rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+};
+```
+
+Notes: 
+- The client-side version of `GlobalSearchResultProvider` is slightly difference than the 
+server one, as there is no `context` parameter on the `find` signature.
+
+### Plugin API
+
+#### common types
+
+```ts
+/**
+ * Response returned from the {@link GlobalSearchServiceStart | global search service}'s `find` API
+ */
+type GlobalSearchResponse<ResultType extends GlobalSearchResult = GlobalSearchResult> = {
+  /**
+   * Current results fetched from the providers.
+   */
+  results: ResultType[];
+  /**
+   * Is the search complete. Will only be true during the last emission of the `GlobalSearchServiceStart.search` result observable.
+   * A search is considered complete either when all underlying providers completed their observable, or when the timeout is reached.
+   */
+  complete: boolean;
+};
+```
+
+#### server API
+
+```ts
+/** @public */
+interface GlobalSearchServiceSetup {
+  registerResultProvider(provider: GlobalSearchResultProvider);
+}
+
+/** @public */
+interface GlobalSearchServiceStart {
+  find(
+    term: string,
+    options: GlobalSearchOptions,
+    request: KibanaRequest
+  ): Observable<GlobalSearchResponse>;
+}
+```
+
+#### public API
+
+```ts
+/**
+ * Enhanced {@link GlobalSearchResult | result type} for the client-side,
+ * to allow navigating to a given result.
+ */
+interface NavigableGlobalSearchResult extends GlobalSearchResult {
+  /**
+   * Navigate to this result's associated url. If the result is on this kibana instance, user will be redirected to it 
+   * in a SPA friendly way using `application.navigateTo`, else, a full page refresh will be performed.
+   */
+  navigateTo: () => void;
+}
+
+/** @public */
+interface GlobalSearchServiceSetup {
+  registerResultProvider(provider: GlobalSearchResultProvider);
+}
+
+/** @public */
+interface GlobalSearchServiceStart {
+  find(
+    term: string,
+    options: GlobalSearchOptions,
+  ): Observable<GlobalSearchResponse<NavigableGlobalSearchResult>>;
+}
+```
+
+Notes: 
+- The public API is very similar to its server counterpart. The differences are:
+  - The `registerResultProvider` setup APIs share the same signature, however the input `GlobalSearchResultProvider` 
+  types are different on the client and server.
+  - The `find` start API signature got a `KibanaRequest` for `server`, when this parameter is not present for `public`.
+- The `find` API returns a observable of `NavigableGlobalSearchResult` instead of plain `GlobalSearchResult`. This type 
+is here to enhance results with a `navigateTo` method to let `core` handle the navigation logic, which is 
+non-trivial. See the [Redirecting to a result](#redirecting-to-a-result) section for more info.
+
+#### http API
+
+An HTTP API will be exposed on `/api/core/global_search/find` to allow the client-side `GlobalSearchService` 
+to fetch results from the server-side result providers.
+
+It should be very close to:
+
+```ts
+router.post(
+  {
+    path: '/api/core/global_search/find',
```

Comment:
Do we want this to be public and available at `/api`, as opposed to internal and available at `/internal`? Do we have any external consumers in mind?

Reply:
Good point. this is only internal for now, will move to `/internal`. That would also allow having a different public API for external consumers when needed (see https://github.com/elastic/kibana/pull/64284#discussion_r414903332).